src/compiler/transformers/ts.ts

```diff
@@ -1969,7 +1969,16 @@ namespace ts {
          * @param node The type reference node.
          */
         function serializeTypeReferenceNode(node: TypeReferenceNode): SerializedTypeNode {
-            const kind = resolver.getTypeReferenceSerializationKind(node.typeName, currentScope);
+            // node might be a reference to type variable, which can be scoped to a class declaration, in addition to the regular
+            // TypeScript scopes. Walk up the AST to find the next class and use that as the lookup scope.
+            let scope: Node = node;
+            while (scope.parent && scope !== currentScope) {

```

Comment:
We shouldn't be using `.parent` in a transform, as a `before` transform can unset parent pointers. The correct scope should be captured, like `currentScope`, as the transform descends through the tree.

Reply:
Given how `currentScope` is used, I think it is acceptable simply to add the class body to the list of things captured by the transform as a possible `currentScope`, I believe, since it is just used for resolving names (like for this serialization) and checking if we're at the top level of a file.

Reply:
That was my first idea, too. That does however break variable merging in emit, i.e. it adds in a duplicated var declaration:

```
class C {
    constructor() {
        this.foo = 10;
    }
}
C.bar = 20;
var C;  // <-- ADDED
(function (C) {
    C.x = 10;
})(C || (C = {}));
```

I tried debugging, and got lost somewhere in merge declaration markers. plz send help ;-)